primer3/p3helpers.pyx

```diff
@@ -0,0 +1,290 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+'''
+primer3.p3helpers
+~~~~~~~~~~~~~~~~~~~~~~
+
+Contains Cython functions and classes that aid primer design
+
+'''
+from cpython.mem cimport (
+    PyMem_Free,
+    PyMem_Malloc,
+)
+from libc.string cimport memcpy
+
+
+cdef extern from "p3helpers.h":
+    const char COMP_BASE_LUT[128]
+    const char SANITIZE_LUT[128]
+
+
+cdef:
+    char INVALID_BASE = 63 # '?' character
+
+
+cdef inline void copy_c_char_buffer(
+    char*   in_buf,
+    char*   out_buf,
+    int     str_length,
+):
+    '''Copy string to buffer.  Unsafe
+
+    Args:
+        in_buf: input buffer
+        out_buf: output buffer. Assume buffer size is equal to or greater than
+            in_buf size.
+        str_length: known number of bytes to copy
+    '''
+    memcpy(out_buf, in_buf, <size_t> str_length + 1)
+
+
+cdef int ss_rev_comp_seq(char* seq, Py_ssize_t length) nogil:
+    '''Reverse complement sequence C string
+
+    Args:
+        seq: sequence to sanitize
+        length: length of sequence
+
+    Returns:
+        0 on success, 1 on error (Invalid base)
+    '''
+    cdef:
+        char        temp
+        char*       end_ptr = NULL
+
+    end_ptr = seq + (length - 1)
+
+    while end_ptr > seq:
+        temp = end_ptr[0]
+        end_ptr[0] = COMP_BASE_LUT[<unsigned char> seq[0]]
+        if end_ptr[0] == INVALID_BASE:
+            return 1
+        end_ptr -= 1
+        seq[0] = COMP_BASE_LUT[<unsigned char> temp]
+        if seq[0] == INVALID_BASE:
+            return 1
+        seq += 1
+
+
+    if length % 2:
+        seq[0] = COMP_BASE_LUT[<unsigned char> seq[0]]
+
+    return 0
+
+cdef int ss_sanitize_seq(
+    char* seq,
+    Py_ssize_t length,
+) nogil:
+    '''Sanitize sequence C string IUPAC non-{A,C,G,T,a,c,g,t} bases with {N, n}
+
+    Args:
+        seq: sequence to sanitize
+        length: length of sequence
+
+    Returns:
+        0 on success, 1 on error (Invalid base)
+    '''
+    cdef:
+        char*       end_ptr = seq + length
+
+    while seq < end_ptr:
+        seq[0] = SANITIZE_LUT[<unsigned char> seq[0]]
+        if seq[0] == INVALID_BASE:
+            return 1
+        seq += 1
+    return 0
+
+cdef inline void reverse_complement_cb(
+    char* seq,
+    char* out,
+    Py_ssize_t length,
+):
+    copy_c_char_buffer(seq, out, length)
+    ss_rev_comp_seq(out, length)
```

Comment:
We should probably check (or forward along) the return code here? If I'm following the control flow correctly, an invalid base would just cause the reverse comp to cease and you could end up w/ a partially rev comped seq?

Reply:
(although maybe this method isn't used elsewhere? i see that you check the return codes pretty religiously elsewhere)

Reply:
I'm going to yank `reverse_complement_cb` since I didn't mean to leave it in since I didn't test it.